### PR TITLE
MAPI V2 - add resources to change the status of a subscription

### PR DESCRIPTION
**Pull request by phiz71**
## Issue

https://gravitee.atlassian.net/browse/APIM-2006

## Description

You can now close, pause, and resume a subscription with MAPI v2 by calling these resources:
 - POST /apis/{apiId}/subscriptions/{subscriptionId}/_close
 - POST /apis/{apiId}/subscriptions/{subscriptionId}/_pause
 - POST /apis/{apiId}/subscriptions/{subscriptionId}/_resume

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-prtvtcwral.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource.java [hunk 1]
```diff
@@ -289,6 +289,51 @@ public Response updateApiPlan(
             .build();
     }
 
+    @POST
+    @Path("/{subscriptionId}/_close")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Permissions({ @Permission(value = RolePermission.API_SUBSCRIPTION, acls = { RolePermissionAction.UPDATE }) })
+    public Response closeApiPlan(@PathParam("subscriptionId") String subscriptionId) {
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource.java]: `closeApiSubscription`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource.java [hunk 2]
```diff
@@ -289,6 +289,51 @@ public Response updateApiPlan(
...
+        }
+
+        return Response.ok(subscriptionMapper.map(subscriptionService.close(executionContext, subscriptionId))).build();
+    }
+
+    @POST
+    @Path("/{subscriptionId}/_pause")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Permissions({ @Permission(value = RolePermission.API_SUBSCRIPTION, acls = { RolePermissionAction.UPDATE }) })
+    public Response pauseApiPlan(@PathParam("subscriptionId") String subscriptionId) {
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource.java]: `pauseApiSubscription`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource.java [hunk 3]
```diff
@@ -289,6 +289,51 @@ public Response updateApiPlan(
...
+        }
+
+        return Response.ok(subscriptionMapper.map(subscriptionService.pause(executionContext, subscriptionId))).build();
+    }
+
+    @POST
+    @Path("/{subscriptionId}/_resume")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Permissions({ @Permission(value = RolePermission.API_SUBSCRIPTION, acls = { RolePermissionAction.UPDATE }) })
+    public Response resumeApiPlan(@PathParam("subscriptionId") String subscriptionId) {
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource.java]: `resumeApiSubscription`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml [hunk 4]
```diff
@@ -1055,6 +1055,75 @@ paths:
...
+                - API Subscriptions
+            summary: Close one API's subscription
+            description: |-
+                Close the API's subscription.
+
+                User must have the API_SUBSCRIPTION[UPDATE] permission.
+            operationId: closeApiSubscription
+            responses:
+                "200":
+                    description: API's subscription successfully updated
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml]: successfully `closed`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml [hunk 5]
```diff
@@ -1055,6 +1055,75 @@ paths:
...
+                - API Subscriptions
+            summary: Pause one API's subscription
+            description: |-
+                Pause the API's subscription.
+
+                User must have the API_SUBSCRIPTION[UPDATE] permission.
+            operationId: pauseApiSubscription
+            responses:
+                "200":
+                    description: API's subscription successfully updated
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml]: successfully `paused`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml [hunk 6]
```diff
@@ -1055,6 +1055,75 @@ paths:
...
+                - API Subscriptions
+            summary: Resume a paused API's subscription
+            description: |-
+                Resume a previously paused subscription.
+
+                User must have the API_SUBSCRIPTION[UPDATE] permission.
+            operationId: resumeApiSubscription
+            responses:
+                "200":
+                    description: API's subscription successfully updated
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml]: successfully `resumed`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource_PauseTest.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        when(subscriptionService.pause(GraviteeContext.getExecutionContext(), SUBSCRIPTION))
+            .thenReturn(subscriptionEntity.toBuilder().status(SubscriptionStatus.PAUSED).build());
+
+        final Response response = rootTarget().request().post(Entity.json(null));
+        assertEquals(OK_200, response.getStatus());
+
+        var subscription = response.readEntity(Subscription.class);
+        assertEquals(subscriptionEntity.getId(), subscription.getId());
+
+        verify(subscriptionService, times(1)).pause(GraviteeContext.getExecutionContext(), SUBSCRIPTION);
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource_PauseTest.java:125]: `times(1)` is by default if I remember well.

**Review by jhaeyaert — CHANGES_REQUESTED** [review]

**Comment by phiz71** [comment]
Copy/paste issues only 🤦🏻‍♂️.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource_PauseTest.java (hunk 7 above)
Comment (phiz71) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiSubscriptionsResource_PauseTest.java:125]: You're right, but already used in many other tests.
We can do a search&replace later on all the source code.

**Review by jhaeyaert — APPROVED** [review]

**Pull request merged**